tsdb/store.go

```diff
@@ -678,19 +678,25 @@ func (s *Store) Shard(id uint64) *Shard {
 // longer in a "bad" state. This method will return any shards that
 // were removed from the cache.
 func (s *Store) ClearBadShardList() map[uint64]error {
+	s.badShards.mu.Lock()
 	if s.badShards.shardErrors == nil {
-		badShards := make(map[uint64]error)
-		return badShards
+		s.Logger.Warn("badShards was nil")
+		s.badShards.shardErrors = make(map[uint64]error)
 	}
 	badShards := maps.Clone(s.badShards.shardErrors)
 	clear(s.badShards.shardErrors)
+	s.badShards.mu.Unlock()
```

Comment:
I would use `defer` immediately after the `Lock` call as future-proofing against other code paths being added, as well as idiomatic Go.